### ISSUE TITLE
line-breaks not respected for ng-content with CDATA

### ISSUE THREAD

**Issue by muuvmuuv** [angular/angular#42542]
# Bug Report

### Affected Package

 The issue is caused by package @angular/core

### Is this a regression?

 Yes, the previous version in which this bug was not present was: ....

### Description

I am trying to create a code component with line-number and syntax highlighting, unfortunately I cannot get line-breaks working with it. It seems ng-content just strips them. I don't really know if this is a bug or something I have missed, but it seems like one.

## Minimal Reproduction

 https://stackblitz.com/edit/angular-ivy-p3kmma?file=src/app/code/code.component.ts

## Exception or Error
<pre><code>

</code></pre>

## Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 12.0.3
Node: 14.17.0
Package Manager: npm 6.14.13
OS: darwin x64

Angular: 12.0.4
... animations, cdk, common, compiler, compiler-cli, core, forms
... material, platform-browser, platform-browser-dynamic, router
... service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.3
@angular-devkit/build-angular   12.0.3
@angular-devkit/core            12.0.3
@angular-devkit/schematics      12.0.3
@angular/cli                    12.0.3
@schematics/angular             12.0.3
rxjs                            6.6.7
typescript                      4.2.4
</code></pre>

**Anything else relevant?**

**Comment by JoostK** [comment]
You're seeing the effect of whitespace removal as determined by `preserveWhitespaces`, which defaults to `false`. You could set `preserveWhitespaces: true` in the `@Component` decorator, or use the `ngPreserveWhitespaces` attribute to preserve whitespace only in specific parts of the template:

```html
<app-code ngPreserveWhitespaces><![CDATA[<app-tabs [memorize]="'first'">
    <app-tab title="One">
      app-tab 1 content
    </app-tab>
    <app-tab title="<i>Two</i>">
      app-tab 2 content with html title
    </app-tab>
    <app-tab title="Three">
      app-tab 3 content
    </app-tab>
  </app-tabs>]]>
</app-code>
```

That said, I agree that the whitespace removal pass should not touch CDATA texts, similar to #26276 which targets whitespace in expression literals.

**Comment by muuvmuuv** [comment]
Hey @JoostK thanks for the fast response. I am already using `preserveWhitespaces: true` or do you mean that needs to be set in `app.component.ts`? I think setting it in the place where I *use* a component (same with using `ngPreserveWhitespaces`) isn't really user-friendly and could be missed by some other dev not checking the visual output while just creating a docs page e.g.

Anyway, I will add it to the components notes to use `ngPreserveWhitespaces`. But maybe we could extend `ngPreserveWhitespaces` with more values like `boolean | "trim" | "trimr" | "triml"` which would make the HTML also more readable:

```html
<app-code><![CDATA[
  <app-tabs [memorize]="'first'">
    <app-tab title="One">
      app-tab 1 content
    </app-tab>
    <app-tab title="<i>Two</i>">
      app-tab 2 content with html title
    </app-tab>
    <app-tab title="Three">
      app-tab 3 content
    </app-tab>
  </app-tabs>
]]></app-code>
```

instead of:

```html
<app-code ngPreserveWhitespaces><![CDATA[<app-tabs [memorize]="'first'">
  <app-tab title="One">
    app-tab 1 content
  </app-tab>
  <app-tab title="<i>Two</i>">
    app-tab 2 content with html title
  </app-tab>
  <app-tab title="Three">
    app-tab 3 content
  </app-tab>
</app-tabs>]]></app-code>
```

**Comment by JoostK** [comment]
>  I am already using preserveWhitespaces: true or do you mean that needs to be set in app.component.ts? I think setting it in the place where I use a component (same with using ngPreserveWhitespaces) isn't really user-friendly and could be missed by some other dev not checking the visual output while just creating a docs page e.g.

Yes, it needs to be in `app.component.ts` as that's where the content is declared.

> Anyway, I will add it to the components notes to use ngPreserveWhitespaces. But maybe we could extend ngPreserveWhitespaces with more values like boolean | "trim" | "trimr" | "triml" which would make the HTML also more readable:

That wouldn't work, as CDATA should be left untouched. You'd then at least need:

```html
<app-code ngPreserveWhitespaces>
  <![CDATA[<app-tabs [memorize]="'first'">
  <app-tab title="One">
    app-tab 1 content
  </app-tab>
  <app-tab title="<i>Two</i>">
    app-tab 2 content with html title
  </app-tab>
  <app-tab title="Three">
    app-tab 3 content
  </app-tab>
</app-tabs>]]>
</app-code>
```

But `ngPreserveWhitespaces` should only be considered a workaround here; it shouldn't be needed to preserve the CDATA content (as per this bug report)
